### PR TITLE
The KubernetesPodOperator Comprehensive Guide

### LINKED ISSUE #8970: Improve guide for KubernetesPodOperator
Hello,

A better guide that describes how to use KubernetesPodOperator would be useful. This is one of the most frequently used Airflow operators, so many users have problems with this operator. 
https://airflow.readthedocs.io/en/latest/howto/operator/kubernetes.html
We currently have a guide that does not describe all the basic concepts.  It is not very useful for the end user.

It would be fantastic if the following topics were discussed. 

- [x] How does KubernetesPodOperator work?  Send request via Kubernetes API and waiting for the end. Airlfow uses labels to identity pod.
- [ ] How to use cluster ConfigMaps, Secrets, and Volumes with Pod?
- [ ] How to define env?
- [ ] How to use KubernetesPodOperaotr with YAML file/JSON Spec?
- [ ] How to define limit/requests resources?
- [ ] How to use KubernetesPodOperaotrs with Private Docker Image Registry?
- [x] How does XCOM work and how to use it?
- [ ] How to debug problems with this operator? kubectl/kubernetes logs/kubernetes events 
- [ ] How to use KubernetesPodOperator with GKE? GKEStartPodOperator
- [ ] What is the difference between KubernetesPodOperator and KubernetesExecutor?
- [ ] How to deal with common problems? FAQ section
- [ ] Best practice. e.g. separate node pool

If anyone is interested in this task, I am willing to provide all the necessary tips and information. If you are interested in writing a guide, you do not have to describe all the sections. Even one section will be very helpful and easier to learn about Airflow.

If anyone wants to learn more about this operator, I invite you to read the following articles.
https://cloud.google.com/composer/docs/how-to/using/using-kubernetes-pod-operator
https://www.astronomer.io/docs/kubepodoperator/
https://www.astronomer.io/docs/cli-kubepodoperator/
https://medium.com/bluecore-engineering/were-all-using-airflow-wrong-and-how-to-fix-it-a56f14cb0753

Are you wondering how to start contributing to this project? Start by reading our [contributor guide](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst)

### PR DESCRIPTION
**Pull request by drdenzy**
The guide aims to alleviate most of the common challenges users
encounter when using KubernetesPodOperator. It explains the underlying
concept of the operator. And It also contains several 'HOWTOs' and best
practices for using the operator.

Ultimately, this guide will help users gain better understanding of how
Airflow tasks are executed in pods in kubernetes clusters.

closes: #8970

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: docs/apache-airflow-providers-cncf-kubernetes/operators.rst [hunk 1]
```diff
@@ -73,14 +103,21 @@ and type safety. While we have removed almost all Kubernetes convenience classes
...
-a substitute for a Kubernetes object spec definition that is able
-to be run in the Airflow scheduler in the DAG context. If using the operator, there is no need to create the
-equivalent YAML/JSON object spec for the Pod you would like to run.
-The YAML file can still be provided with the ``pod_template_file`` or even the Pod Spec constructed in Python via
-the ``full_pod_spec`` parameter which requires a Kubernetes ``V1Pod``.
+
+How to use KubernetesPodOperator with YAML file/JSON spec?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * WIP
```
Comment (XD-DENG) [inline docs/apache-airflow-providers-cncf-kubernetes/operators.rst:108]: Does these `WIP`s mean this PR is still a draft at this moment? If so, let's convert it to draft.

Comment (drdenzy) [reply docs/apache-airflow-providers-cncf-kubernetes/operators.rst:108]: Yes. It is more of a draft now - lots more write-ups will be added.

File: docs/apache-airflow-providers-cncf-kubernetes/operators.rst [hunk 2]
```diff
@@ -35,23 +36,52 @@ you to create and run Pods on a Kubernetes cluster.
...
+
+* The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` enables task-level
+  resource configuration and is optimal for custom Python dependencies that are not available through the
+  public PyPI repository.
+
+* It allows users to supply a template YAML file using the ``pod_template_file`` parameter.
+
+* It allows isolation of deployments, configuration reuse, delegation and better management of secrets.
+
+* Ultimately, it allows Airflow to act a job orchestrator - no matter the language those jobs are written in.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-cncf-kubernetes/operators.rst:71]: I think a nice succinct way to think of it is: 'Easy way to run any image on Kubernetes as a task', and I feel like this should be first! (related to the orchestrator and any language points imo)

File: docs/apache-airflow-providers-cncf-kubernetes/operators.rst [hunk 3]
```diff
@@ -35,23 +36,52 @@ you to create and run Pods on a Kubernetes cluster.
...
 The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` uses the
 Kubernetes API to launch a pod in a Kubernetes cluster. By supplying an
 image URL and a command with optional arguments, the operator uses the Kube Python Client to generate a Kubernetes API
 request that dynamically launches those individual pods.
+Under the hood, :class:`~airflow.providers.cncf.kubernetes.hooks.kubernetes.KubernetesHook` creates the connection to
+the Kubernetes API server.
+Essentially, KubernetesPodOperator packages all the supplied parameters into a request object which is then shipped off
+to Kubernetes API Server so that the pod to execute your task is created. Whenever a task is triggered, a new worker pod
+is spun up to execute that task. And once the task is completed, by default the worker pod is deleted
+and the resources reclaimed.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-cncf-kubernetes/operators.rst:49]: I'd be careful with terminology here. "Worker pods" are a KubeExecutor thing, not KPO.

File: docs/apache-airflow-providers-cncf-kubernetes/operators.rst [hunk 4]
```diff
@@ -19,11 +19,12 @@
...
 
-KubernetesPodOperator
-=====================
+KubernetesPodOperator - The Comprehensive Guide
+===============================================
 
 The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` allows
-you to create and run Pods on a Kubernetes cluster.
+you to create and run Pods on a Kubernetes cluster. The task wrapped in the KubernetesPodOperator is then executed in
+these pods.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-cncf-kubernetes/operators.rst:27]:
```suggestion
you to create and run a Pod on a Kubernetes cluster as a task.
```

Maybe?

File: docs/apache-airflow-providers-cncf-kubernetes/operators.rst [hunk 5]
```diff
@@ -35,23 +36,52 @@ you to create and run Pods on a Kubernetes cluster.
...
-
 How does this operator work?
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
 The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` uses the
 Kubernetes API to launch a pod in a Kubernetes cluster. By supplying an
 image URL and a command with optional arguments, the operator uses the Kube Python Client to generate a Kubernetes API
 request that dynamically launches those individual pods.
+Under the hood, :class:`~airflow.providers.cncf.kubernetes.hooks.kubernetes.KubernetesHook` creates the connection to
+the Kubernetes API server.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-cncf-kubernetes/operators.rst]:
```suggestion
```

I think this is redundant with the beginning of the this paragraph.

File: docs/apache-airflow-providers-cncf-kubernetes/connections/kubernetes.rst [hunk 6]
```diff
@@ -20,7 +20,9 @@
 Kubernetes cluster Connection
 =============================
 
-The Kubernetes cluster Connection type enables connection to a Kubernetes cluster by :class:`~airflow.providers.cncf.kubernetes.operators.spark_kubernetes.SparkKubernetesOperator` tasks. They are not used by ``KubernetesPodOperator`` tasks.
+The Kubernetes cluster Connection type enables connection to a Kubernetes cluster by
+:class:`~airflow.providers.cncf.kubernetes.operators.spark_kubernetes.SparkKubernetesOperator` tasks.
+They are not used by ``KubernetesPodOperator`` tasks.
```
Comment (jedcunningham) [inline docs/apache-airflow-providers-cncf-kubernetes/connections/kubernetes.rst]:
```suggestion
They are not used by :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator tasks.
```

Probably worth linking it.

**Comment by eladkal** [comment]
Does this PR aims to enhance/replace https://github.com/apache/airflow/pull/13405 ?

**Comment by Dr-Denzy** [comment]
> Does this PR aim to enhance/replace #13405 ?

The aim is to enhance #13405 in a way to provides a comprehensive doc for users.

**Pull request closed without merging**